### DISCUSSION TITLE
Files are suddenly stripped from the Master Branch - How do we trace the entire List ?

### DISCUSSION THREAD

**Discussion by ppmgitrepo** [community/community#52813] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

Suddenly a large number of files that had versions on the Master branch have got delinked from the Master branch. 
When we access the file directly via the Tag URL it shows the file , but displays this error 

```This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.```

Have 3 issues now. 
1. A git Log is not showing any activity that might have caused this for a specific checkin tag
2. How do we get a complete listing of such files ?
3. What is shortest route the push these files back into the Master branch via a checkin ?

Additional Info 
Say a file with a commit tag ```[a5ae00d]``` earlier (5 days ago) belong to the Master Branch, but now suddenly for the same commit tag ```[a5ae00d]``` , it shows as no longer part of the Master branch. 

You can still access the file via the git URL 
https://github.com/<repo>/xxx/blob/```a5ae00d```/sub-path/filename

Thanks
Czd

**Comment by airtower-luna** [comment]
It sounds like someone force-pushed the branch. That won't be visible in `git log` because a force-push *changes the history* which you see in the log.

> What is shortest route the push these files back into the Master branch via a checkin ?

That really depends on the state you want to achieve on the branch. Do you want to combine it with the tag you mentioned? Do you want to reset it to how it is at the tag? It would be ideal if you still have a local copy of the repository where the branch looks like you want it to (before the incident), in that case you could just force-push that over the current state. But double check it's really what you want before any force-pushing or you might end up making the situation worse. :warning: 

Either way I'd recommend setting up a [branch protection rule](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches) that prohibits force-pushes to prevent future incidents of this kind (after you fix the branch).

**Reply by ppmgitrepo** [reply]
@airtower-luna 

Thanks for your response. 
I have a query will the following set of commands pull the orphaned version of the file back into the Master ?
(Linux OS CLI) 

```
git clone <repo>
git checkout Master
git pull origin Master
git merge a5ae00d:<path>/<file>
git push origin Master

```

Since the repository is huge , I need to be absolutely certain that I will not corrupt the repository

**Reply by ppmgitrepo** [reply]
i.e. to add to my prev question 
I do not want any automatic merge to happen for any file other than the ones I am specifying by name+path
So my doubt was if this command will process only 1 single file ?
`git merge a5ae00d:<path>/<file>
`

**Reply by airtower-luna** [reply]
> I do not want any automatic merge to happen for any file other than the ones I am specifying by name+path
> So my doubt was if this command will process only 1 single file ?

You can't merge a single file, a merge will always merge the whole repository state.

If you want to set just one file to the state at a past commit, you can [checkout the file as it was at that commit](https://git-scm.com/docs/git-checkout#Documentation/git-checkout.txt-emgitcheckoutem-f--ours--theirs-m--conflictltstylegtlttree-ishgt--ltpathspecgt82308203), e.g.:

```
git checkout a5ae00d -- path/to/your/file
```

This will overwrite `path/to/your/file`, if you have any changes you want to keep stash or save them elsewhere first. :warning: The checkout will *only* change the file on disk, after that you can add & commit as usual.

**Reply by ppmgitrepo** [reply]
This one I tired and it did not work .. you can no longer check it out via CLI 

```
$ git checkout 065aa24 -- Pascal_OLTP/geog/po/12.0.0/admin/sql/GEPSPOVENDORSUBSCRIBE11I_PKG.pkb
fatal: invalid reference: 065aa24

```
However you can manually download the file via the Github default Webpage.
Is that the only way to restore it ?

**Reply by airtower-luna** [reply]
The error message means the commit ID you tried to use is invalid. Make sure it's correct (a commit you actually have on disk), and use the full ID (the SHA-1 hash), not a shortened version. If you have a tag or branch that points at the correct commit that works, too.

If the commit doesn't actually exist on your disk (only in a remote repository) *and* isn't reachable from any branch or tag you could fetch then yes, downloading the file will be your only option.

**Reply by ppmgitrepo** [reply]
The commit id is correct 
using the URL you can access the ame commit id. 
i.e. 
`https://github.com/xxxYOURORGxxxx/Pascal_OLTP/blob/065aa24/Pascal_OLTP/geog/po/12.0.0/admin/sql/GEPSPOVENDORSUBSCRIBE11I_PKG.pkb`

When we load it via this it displays but says 
"This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository."
which circles back to the original problem reported. 

The only way to fix it is from the URL do a manual raw download 
and then check it in. 
As via CLI am not able to download it.

**Reply by airtower-luna** [reply]
> The only way to fix it is from the URL do a manual raw download
> and then check it in.

In that situation a manual download is probably the only way, yes.

> "This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository."

That means the commit is *not* part of the history of any branch or tag in your repository. As the message says it *may* be from a fork. It could also be that the commit was pushed to your repository as part of a branch, and someone then force-pushed a version of that branch that didn't include the commit, making it unreachable. I can't say what actually happened.
